pinot-core/src/main/java/org/apache/pinot/core/query/aggregation/groupby/DictionaryBasedGroupKeyGenerator.java

```diff
@@ -831,6 +777,156 @@ private String getGroupKey(IntArray rawKey) {
+     * upper bound is not reached.
+     */
+    public int getGroupId(int rawKey, int groupIdUpperBound) {
+      // NOTE: Key 0 is reserved as the null key. Use (rawKey + 1) as the internal key because rawKey can never be -1.
+      int internalKey = rawKey + 1;
+      int index = (HashCommon.mix(internalKey) << 1) & _mask;
+      int key = _keyValueHolder[index];
+
+      // Handle hash hit separately for better performance
+      if (key == internalKey) {
+        return _keyValueHolder[index + 1];
+      }
+      if (key == 0) {
+        return _size < groupIdUpperBound ? addNewGroup(internalKey, index) : INVALID_ID;
+      }
+
+      // Hash collision
+      while (true) {
+        index = (index + 2) & _mask;
+        key = _keyValueHolder[index];
+        if (key == internalKey) {
+          return _keyValueHolder[index + 1];
+        }
+        if (key == 0) {
+          return _size < groupIdUpperBound ? addNewGroup(internalKey, index) : INVALID_ID;
+        }
+      }
+    }
+
+    private int addNewGroup(int internalKey, int index) {
+      int groupId = _size++;
+      _keyValueHolder[index] = internalKey;
+      _keyValueHolder[index + 1] = groupId;
+      if (_size > _maxNumEntries) {
+        expand();
+      }
+      return groupId;
+    }
+
+    private void expand() {
+      _capacity <<= 1;
+      int holderSize = _capacity << 1;
+      int[] oldKeyValueHolder = _keyValueHolder;
+      _keyValueHolder = new int[holderSize];
+      _mask = holderSize - 1;
+      _maxNumEntries <<= 1;
+      int oldIndex = 0;
+      for (int i = 0; i < _size; i++) {
+        while (oldKeyValueHolder[oldIndex] == 0) {
+          oldIndex += 2;
+        }
+        int key = oldKeyValueHolder[oldIndex];
+        int value = oldKeyValueHolder[oldIndex + 1];
+        int newIndex = (HashCommon.mix(key) << 1) & _mask;
+        if (_keyValueHolder[newIndex] != 0) {
+          do {
+            newIndex = (newIndex + 2) & _mask;
+          } while (_keyValueHolder[newIndex] != 0);
+        }
+        _keyValueHolder[newIndex] = key;
+        _keyValueHolder[newIndex + 1] = value;
+        oldIndex += 2;
+      }
+    }
+
+    public Iterator<Entry> iterator() {
+      return new Iterator<Entry>() {
+        private final Entry _entry = new Entry();
+        private int _index;
+        private int _numRemainingEntries = _size;
+
+        @Override
+        public boolean hasNext() {
+          return _numRemainingEntries > 0;
+        }
+
+        @Override
+        public Entry next() {
+          int key;
+          while ((key = _keyValueHolder[_index]) == 0) {
+            _index += 2;
+          }
+          _entry._rawKey = key - 1;
+          _entry._groupId = _keyValueHolder[_index + 1];
+          _index += 2;
+          _numRemainingEntries--;
+          return _entry;
+        }
+
+        @Override
+        public void remove() {
+          throw new UnsupportedOperationException();
+        }
+      };
+    }
+
+    /**
```

Comment:
There is no trimming here like the way it happens in combiner where we sort and trim for. We are just zeroing out the keyValueHolder array and allocating a new array if size > MAX_CACHING_MAP_SIZE

Reply:
It is basically clear + trim if the size is larger than the threshold